libs/etcd_broker/src/lib.rs

```diff
@@ -282,54 +290,64 @@ pub async fn subscribe_to_safekeeper_timeline_updates(
     })
 }
 
+fn extract_key_value_str(kv: &KeyValue) -> Result<(&str, &str), BrokerError> {
+    let key = kv.key_str().map_err(|e| {
+        BrokerError::EtcdClient(e, "Failed to extract key str out of etcd KV".to_string())
+    })?;
+    let value = kv.value_str().map_err(|e| {
+        BrokerError::EtcdClient(e, "Failed to extract value str out of etcd KV".to_string())
+    })?;
+    Ok((key, value))
+}
+
 fn parse_etcd_key_value(
     subscription_kind: SubscriptionKind,
     regex: &Regex,
-    kv: &KeyValue,
-) -> Result<Option<(ZTenantTimelineId, SafekeeperTimeline)>, BrokerError> {
-    let caps = if let Some(caps) = regex.captures(kv.key_str().map_err(|e| {
-        BrokerError::EtcdClient(e, format!("Failed to represent kv {kv:?} as key str"))
-    })?) {
-        caps
-    } else {
-        return Ok(None);
+    key_str: &str,
+    value_str: &str,
+) -> Result<(ZTenantTimelineId, SafekeeperTimeline), BrokerError> {
+    let key_captures = match regex.captures(key_str) {
+        Some(captures) => captures,
+        None => {
+            return Err(BrokerError::ParsingError(format!(
+                "KV has unexpected key '{key_str}' that does not match required regex {regex}"
+            )));
+        }
     };
+    let info = serde_json::from_str(value_str).map_err(|e| {
+        BrokerError::ParsingError(format!(
+            "Failed to parse '{value_str}' as safekeeper timeline info: {e}"
+        ))
+    })?;
 
     let (zttid, safekeeper_id) = match subscription_kind {
```

Comment:
Key is the same in all kinds of subs, it would be simpler to avoid this match. (and generally 3 kinds of regexp I guess)

Reply:
That required splitting off the etcd_prefix first, but done.